String interpolators for codegens

Lewix opened the pull request:
The changes in this pull request add two string interpolators, `gen` and `src`, which simplify writing code generation cases. Both of the interpolators automate:
- Quoting of Expressions
- Remapping of Manifests
- Concatenating sequences with commas

On top of this, `gen` calls `stream.println` on the strings and `emitBlock` on any blocks. The result is that code such as:

```
stream.println("val " + quote(sym) + " = { ")
emitBlock(b)
stream.println(quote(getBlockResult(b)))
stream.println("}")
```

Now looks like this:

```
gen"""val $sym = {
     |$b
     |${getBlockResult(b)}
     |}"""
```

Lewix committed:
Replace string concatenation with interpolation

In all of the *Ops.scala files.  Some instances of string.format(...) were also
replaced.

Lewix committed:
Add a string interpolator for code generators

The interpolator will automatically call quote or remap depending on the type of
each argument.  If the argument is already a string, it doesn't call anything.

Lewix committed:
Quote Seqs in the string interpolator

This avoids expressions like 'x.map(quote).mkString(",")' which are hard to
read.

Lewix committed:
Add a string interpolator to do stream.prinltn

Now src"..." simply does the quoting, whereas gen"..." quotes and prints to the
stream.

Lewix committed:
Generate blocks with the gen string interpolator

A subclass of CodegenHelper, contained in GenericNestedCodegen
detects whether to print to the stream or simply call emitBlock.

Stripping the margin becomes more complicated as the context part may
contain a |.

Lewix committed:
Change .check files so the gen code looks better

'emitBlock' prints a new line at the end of the block, so a blank line
is currently generated after every block.  To avoid this one could
remove the new line in the string interpolation, but that's ugly.

vjovanov commented on test-out/epfl/test11-hmm1.check:
```diff
@@ -28,6 +28,7 @@ val x22 = x0(x18)
 val x23 = x21 * x22
 val x24 = x19 + x23
 val x25 = x1(3) = x24
+
```
Why do we get these extra lines for ArrayUpdate and not in other places?

Lewix replied:
The new lines are due to blocks being followed by a newline in the code. So something like:

```
gen"""// code
     |$block
     |${getBlockResult(body)}
     |// more code"""
```

Will call `emitBlock(block)`, print a newline to `stream` and then process `getBlockResult(body)`. This could be avoided, but since it's generated code that nobody is going to read it probably isn't worth it.

vjovanov replied:
Good. I though it is more severe than that.

vjovanov commented on src/common/Equal.scala:
```diff
@@ -96,9 +96,9 @@ trait CLikeGenEqual extends CLikeGenBase {
       rhs match {
         case Equal(a,b) =>
-          emitValDef(sym, quote(a) + " == " + quote(b))
+          emitValDef(sym, raw"${quote(a)} == ${quote(b)}")
```
Why not use src here?

Lewix replied:
Oh yeah, I must have missed some when changing raw to gen/src.

vjovanov commented:
Why do we mix src and raw? I guess src should be used in every place.

Lewix committed:
Replace raw with src in Equal.scala

vjovanov commented:
LGTM.

vjovanov commented:
@tiarkrompf Can we merge this?

TiarkRompf commented:
One thing I want to make sure is that we are not going to quadratic complexity by repeatedly copying generated strings from nested blocks. How is this handled? 

It also seems `emitBlock` is called implicitly, but `getBlockResult` is still explicit. I would suggest to swap it. `emitBlock` is a recursive call, and I think we should keep those visible.

TiarkRompf commented on src/common/IfThenElse.scala:
```diff
@@ -318,7 +318,7 @@ trait CudaGenIfThenElseFat extends CudaGenIfThenElse with CudaGenFat with BaseGe
   import IR._
 
   override def emitFatNode(symList: List[Sym[Any]], rhs: FatDef) = rhs match {
-    case SimpleFatIfThenElse(c,a,b) => sys.error("TODO: implement fat if CUDA codegen")
```
search and replace errors!

Lewix committed:
Fix search and replace errors

Lewix commented:
No, the strings are not copied when generating nested blocks. A string such as:

```
// context
$block
// more context
```

Gets broken into a list of context parts (`["// context\n", "\n// more context"]`) and a list of arguments (`[block]`). The `gen` string interpolator will interleave printing the contexts and the arguments to `stream`, as you would expect. However if it encounters an `Block[_]` argument, instead of calling `stream.println`, it simply calls `emitBlock`. This way, the `emitBlock` method does the printing rather than `gen` having to copy the string produced by `emitBlock` into the stream.

The code where this is handled is in [`GenericNestedCodeGen`](https://github.com/TiarkRompf/virtualization-lms-core/pull/69/files#diff-6dbe6e1943d843b7d0382aaa9e98700eR233) and [`GenericCodeGen`](https://github.com/TiarkRompf/virtualization-lms-core/pull/69/files#diff-6dbe6e1943d843b7d0382aaa9e98700eR200).

I agree it would be better to have `emitBlock` explicit and `getBlockResult` implicit, but I'm not sure if that's possible. I would need `emitBlock(block)` to be passed to the string evaluator by name, but I get a "no by-name parameter type allowed here" error if I try to.

TiarkRompf commented:
Thanks for clarifying -- sounds good. 

For `emitBlock`, maybe we can do something like this:

   ${nestedBlock(block}}

And have `nestedBlock` create a thunk, i.e. a dummy object that wraps the block, and then call `emitBlock` as before when we process that object later.

Lewix commented:
Ok, how about using a case class called `NestedBlock`. Then the code becomes:

```
gen"""val $sym = {
     |${NestedBlock(b)}
     |$b
     |}"""
```

TiarkRompf commented:
I have a slight preference for a method as the public API for reasons of modularity: you can override a method in a subtrait, but you cannot override a case class.

Lewix committed:
Make calls to emitBlock explicit

Added a "nestedBlock" method to do this.  Calls to getBlockResult are now
implicit.

Lewix commented:
I just noticed a missing `$` sign in [`ArrayOps.scala`](https://github.com/TiarkRompf/virtualization-lms-core/pull/69/files#diff-a4c1fd604b706b03b56b3f1d469eaf6bR237) due to my last commit. It should say `out(i) = $blk`. The timing is unfortunate, so would you like me to submit another PR? It might be simpler for you to just add the `$` yourself. Sorry about this.

vjovanov commented:
Just submit another pull request as that is the usual policy.
